### ISSUE TITLE
cancelTransaction() causes REALM_CHANGED/LOCAL_COMMIT missed.

### ISSUE THREAD

**Issue by beeender** [realm/realm-java#3002]
In normal cases pair `Realm.beginTransaction()`/`Realm.commitTransaction()` will trigger realm updates for local thread from `Realm.commitTransaction()`.

But for `Realm.beginTransaction()`/`Realm.cancelTransaction()`, the update event won't be sent anymore. This is not correct and will cause some strange issues since `beginTransaction()` will do advance read and update the local realm instance.

**Comment by cmelchior** [comment]
Nice catch

**Comment by nhachicha** [comment]
There are 4 Use Cases I can see:

## UC 1: Sync _RealmResults_
```java
realm.beginTransaction();
Dog dog1 = realm.createObject(Dog.class);
dog1.setName("Dog1");
Dog dog2 = realm.createObject(Dog.class);
dog2.setName("Dog2");
realm.commitTransaction();


final RealmResults<Dog> all = realm.where(Dog.class).findAll();
all.addChangeListener(new RealmChangeListener<RealmResults<Dog>>() {
    @Override
    public void onChange(RealmResults<Dog> element) {
        assertEquals(2, element.size());// ==> This is not triggered
    }
});

realm.beginTransaction();
dog1.deleteFromRealm();
realm.cancelTransaction();

assertEquals(2, all.size());
// ==> [0] is Invalid Object at this time 
// ==> Registered listener is not triggered
```
This UC could be solved by rerunning the query (invoking `syncIfNeeded`)

## UC 2: Async _RealmResults_
```java
realm.beginTransaction();
Dog dog1 = realm.createObject(Dog.class);
dog1.setName("Dog1");
Dog dog2 = realm.createObject(Dog.class);
dog2.setName("Dog2");
realm.commitTransaction();

final RealmResults<Dog> all = realm.where(Dog.class).findAllAsync();
all.addChangeListener(new RealmChangeListener<RealmResults<Dog>>() {
    @Override
    public void onChange(RealmResults<Dog> element) {
        assertEquals(2, element.size());// ==> Triggered when the query completes (after next Looper event)
    }
});

realm.beginTransaction();
dog1.deleteFromRealm();
realm.cancelTransaction();

assertEquals(0, all.size());
// ==> size is 0, deleted RealmObject doesn't impact the Async query
// ==> Registered listener is triggered

```
## UC 3: Sync _RealmObject_
```java
realm.beginTransaction();
Dog dog1 = realm.createObject(Dog.class);
dog1.setName("Dog1");
Dog dog2 = realm.createObject(Dog.class);
dog2.setName("Dog2");
realm.commitTransaction();

dog1.addChangeListener(new RealmChangeListener<RealmModel>() {
    @Override
    public void onChange(RealmModel element) {
        assertNotNull(element);// ==> Not triggered 
    }
});

realm.beginTransaction();
dog1.deleteFromRealm();
realm.cancelTransaction();

assertFalse(dog1.isValid());// ==> This should be valid 
```
## UC 4: Async _RealmObject_
```java
realm.beginTransaction();
Dog dog1 = realm.createObject(Dog.class);
dog1.setName("Dog1");
Dog dog2 = realm.createObject(Dog.class);
dog2.setName("Dog2");
realm.commitTransaction();

Dog dog = realm.where(Dog.class).findFirstAsync();

dog.addChangeListener(new RealmChangeListener<RealmModel>() {
    @Override
    public void onChange(RealmModel element) {
        assertNotNull(element);// ==> Triggered with "Dog1"
    }
});

realm.beginTransaction();
dog1.deleteFromRealm();
realm.cancelTransaction();
```

## Work around ?
UC1 could be solved by invoking `syncIfNeeded` we don't have a problem with UC2. The problem with UC3 & partially with UC4 is as long as we delete an object we set the row to `proxy.realmGet$proxyState().setRow$realm(InvalidRow.INSTANCE);` we lost the accessor of the row. This could be either documented or we should mark/tombstone a deleted `RealmObject` until we `commit` the transaction or set back the previously row accessor when we cancel the transaction. WDYT?

**Comment by beeender** [comment]
This will probably fixed by #3834 . But need to add more test case to close this issue.

**Comment by beeender** [comment]
#3834 is merged to master. will be fixed in `v2.4.0`.

**Comment by Zhuinden** [comment]
If `beginTransaction()` causes immediate `RealmChangeListener` call, then local commits can no longer be "missed" in non-commit state.
